---
title: "Gene set enrichment analysis"
author: "Kim Dill-McFarland, <kadm@uw.edu>"
date: "version `r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
---
```{r include=FALSE}
knitr::opts_chunk$set(fig.width = 8.5, fig.height = 3.5)
```

# Overview
In this workshop, we introduce gene set analysis relevant to RNA-sequencing data. In it, we cover:

* Broad Molecular Signatures Database (MSigDB) gene sets
* hypergeometric enrichment with `clusterProfiler`
* gene set enrichment analysis (GSEA) with `fgsea`

During the workshop, we will build an R script together, which will be posted as 'live_notes' after the workshop [here](https://github.com/hawn-lab/workshops_UW_Seattle/tree/master/2021.07.14_GSEA).

# Prior to the workshop

Please [install R](http://www.r-project.org), [RStudio](https://www.rstudio.com/products/rstudio/download/), and the following packages. See the [setup instructions](notes/introR_setup.html) for more details.

```{r message=FALSE, warning=FALSE}
#Data manipulation
library(tidyverse)
#Broad gene set database
library(msigdbr)
#Hypergeo enrichment
library(clusterProfiler)
#GSEA
library(fgsea)
```

# Load data

```{r}
#RNAseq expression data
load("data/RSTR_data_clean_subset.RData")
#Linear model results
model.results <- read_csv("data/RSTR.Mtb.model.results.anno.csv") %>% 
  #Filter results for Mtb vs media condition
  filter(variable == "conditionTB")
```

# Gene sets

Definition: two or more genes with shared function, structure, localization, or any other defined grouping variable

Use: glean interpretable, biologically-relevant meaning from a list of genes of interest

Gene sets can be as general as "cytosolic" to as specific as "up-regulated in blood vessel cells in response to wound in *Roy et al 2007*". They can be defined by an individual research study (as in the latter) up to generally agreed upon knowledge formed from an entire body of literature (like the former). They also vary in size (2 to thousands), are not exclusive (one gene is in many gene sets), and many are redundant (transmembrane, membrane, lipid bilayer...).

It is up to you, the researcher, to select what group of gene sets you want to test. More on this below. Keep in mind, this is definitely not a case for "more is better". For example, getting 200 significant gene sets from your 250 significant genes does not help you much.

## Broad MSigDB

The Broad Institute has collated many useful gene sets in their Molecular Signatures Database ([MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp)). This includes the following used in this workshop (descriptions modified from Broad).

* Hallmark: summarize and represent specific well-defined biological processes. Generated by a computational methodology based on identifying overlaps between gene sets in other MSigDB collections and retaining genes that display coordinate expression
    - Total: 50
    - Examples: glycolysis, inflammatory response, apoptosis
* Curated (C2)
    - Canonical pathways: from pathway databases including [KEGG](https://www.genome.jp/kegg/), [REACTOME](https://reactome.org/), etc. Canonical representations of a biological process compiled by domain experts
    - Total: 2922
    - Examples: Caspase pathway, signaling by NOTCH1, DNA repair
* Gene ontology [GO](http://geneontology.org/) (C5)
    - Biological process: molecular-level activities performed by gene products
    - Total: 7481
    - Examples: viral life cycle, vitamin D biosynthetic process, mitochondrial calcium-ion transmembrane transport

Hallmark is the most general group and is usually where we start our analyses. It helps reduce the complexity of a large gene list down to easily understood terms. It also provides some context when looking at more specific gene sets as you can often group the other sets under a more general Hallmark term. However, because Hallmark is the most general, it is also the most incomplete with 50 gene sets representing just under 4400 genes.

Curated and gene ontology sets are more specific than Hallmark, though they contain a range of levels. They are particularly useful in adding details to a Hallmark term or defining testable hypotheses for follow-up experiments in the lab. They contain a lot more genes than Hallmark but many sets are redundant or overlapping. If there are many significant Hallmark terms, we often do not run these more specific sets as it gives too many significant results for interpretation. 

There are several other sets in MSigDB that are not described here. Some of these are too specific to be of use in our global RNA-sequencing analyses (C2 chemical and genetic perturbations or C7 immunologic signatures) while other covers areas not relevant to our experimental design (C1 chromosome position, C3 gene regulation, C4/C6 cancer-oriented). 

Next we will use gene sets to unravel how monocytes respond to *M. tuberculosis* infection.

# Hypergeometric enrichment

This analysis is a simple enrichment. We ask if our list of significant genes is enriched in gene sets more than random chance would allow.

### Get gene set data

First, we extract the lists of genes in Broad terms from the package `msigdbr`. We format them as a data frames so we can use the `tidyverse`.

```{r}
#Hallmark
H <- as.data.frame(msigdbr(species = "Homo sapiens", 
                           category = "H"))
#Canonical pathways
#Note that we have to call each database and combine them together
C2.CP <- as.data.frame(msigdbr(species = "Homo sapiens", 
                               category = "C2",
                               subcategory = "CP:BIOCARTA")) %>% 
        bind_rows(as.data.frame(msigdbr(species = "Homo sapiens", 
                                category = "C2",
                                subcategory = "CP:KEGG"))) %>% 
        bind_rows(as.data.frame(msigdbr(species = "Homo sapiens", 
                                category = "C2",
                                subcategory = "CP:PID"))) %>% 
        bind_rows(as.data.frame(msigdbr(species = "Homo sapiens", 
                                category = "C2",
                                subcategory = "CP:REACTOME")))
#Gene ontology biological process
C5.BP <- as.data.frame(msigdbr(species = "Homo sapiens", 
                               category = "C5", subcategory = "GO:BP"))
```

Here, we see the start of the Hallmark data.

```{r}
head(H)
```

### Define significant genes

Next, we must define our "significant genes". This type of enrichment is binary in that it only considers if a gene is significant or not. There is no granularity beyond yes/no. Thus, it is very important to consider our FDR cutoff as it is an arbitrary value that can dramatically impact results. 

Let's plot our FDR values as a histogram to help us determine a cutoff. We see that the data are heavily skewed. This is not uncommon when you have a large biological perturbation like bacterial infection.

```{r}
ggplot(model.results, aes(x=FDR)) +
  geom_histogram(bins=100) +
  theme_classic()
```

In fact, there is a large proportion of FDR values that are zero. 

```{r}
table(model.results$FDR == 0)
```

These zeroes are actually a result of limitations in our linear modeling function. After a certain point (in this case, 1E-16), the function stops computing and saves the P-value as zero to save time. 

Here, we will use FDR < 1E-16 to see the genes that change the most with Mtb infection. *This is not necessarily what I'd recommend for these data*. It just for workshop purposes.

```{r}
signif <- model.results %>% 
  filter(FDR <= 1E-16)
```

### Run enrichment

Now we can run the enrichment in `clusterProfiler`. We can use any gene ID in our dataset; it just has to match the gene IDs in the database.

Here, we run the enrichment using ENTREZ ID and the Hallmark database. Note that you must select ONLY the columns needed from the Hallmark database. `enricher( )` isn't smart enough to know which columns to use on its own.

```{r}
#Select unique ENSEMBL IDs for significant genes
signif.entrez <- unique(signif$entrezgene_id)
#Select matching ID column in database
H.entrez <- select(H, gs_name, entrez_gene)
#enrichment
enrich.H <- enricher(gene = signif.entrez, TERM2GENE = H.entrez)
```

Similarly, we could use ENSEMBL IDs.

```{r}
#Select unique ENSEMBL IDs for significant genes
signif.ensembl <- unique(signif$ensembl_gene_id)
#Select matching ID column in database
H.ensembl <- select(H, gs_name, ensembl_gene)
#enrichment
enrich.H <- enricher(gene = signif.ensembl, TERM2GENE = H.ensembl)
```

Importantly, these will yield slightly different results, because the ID databases are not perfect. We must just accept that this is life and choose whichever ID we have / prefer. Since the expression data use ENSEMBL, we'll move forward with that.

```{r}
length(signif.ensembl)
length(signif.entrez)
```

### Extract enrichment results

`clusterProfiler` outputs results in an S4 object. 

```{r}
class(enrich.H)
```

We briefly mentioned this data type in intro R. It works very similarly to S3 (like the expression data) except you use `@` to extract data. For example, the enrichment results are in a data frame.

```{r}
head(enrich.H@result)
```

This includes

```{r}
colnames(enrich.H@result)
```

* ID: Unique name for gene set
* Description: Descriptive text for gene set, often the same as ID
* GeneRatio: Significant genes in gene set / Significant genes in all gene sets tested
* BgRatio: Total genes in gene set / Total genes in all gene sets tested
* pvalue: Significance
* p.adjust: FDR adjust significance
* qvalue: Q-value used in P-value calculation
* geneID: vector of significant gene IDs in gene set. Separated by /
* Count: Significant genes in gene set

These column names aren't the most informative and R doesn't understand the ratios as numbers. 

```{r}
class(enrich.H@result$GeneRatio)
```

So we'll do some additional formatting. We refer to a gene set as a "term" and the Hallmark database as a "category" for brevity.

```{r}
enrich.H.df <- enrich.H@result %>% 
  #separate ratios into 2 columns of data
  separate(BgRatio, into=c("size.term","size.category"), sep="/") %>% 
  separate(GeneRatio, into=c("size.overlap.term", "size.overlap.category"),
               sep="/") %>% 
  #convert to numeric
  mutate_at(vars("size.term","size.category",
                     "size.overlap.term","size.overlap.category"),
                as.numeric) %>% 
  #Calculate k/K
  mutate("k.K"=size.overlap.term/size.term)
```

### Visualize significant enrichment

The most common visualization for enrichment is k/K as it is the ratio of significant genes relative to total genes in the gene set. We also subset to significant enrichment at an defined FDR. Here, let's use FDR < 0.05. This is actually pretty strict of enrichment and people often go up to FDR < 0.2.

```{r}
enrich.H.df %>% 
  filter(p.adjust <= 0.05) %>% 
  #Beautify descriptions by removing _ and HALLMARK
  mutate(Description = gsub("HALLMARK_","", Description),
         Description = gsub("_"," ", Description)) %>% 
  
ggplot(aes(x=reorder(Description, k.K), #Reorder gene sets by k/K values
           y=k.K)) +
  geom_col() +
  theme_classic() +
  #Some more customization to pretty it up
  #Flip x and y so long labels can be read
  coord_flip() +
  #fix labels
  labs(x="Significant genes in set / Total genes in set \nk/K",
       y="Gene set",
       title = "Mtb significant genes (FDR < 1E-16)\nenriched in Hallmark gene sets (FDR < 0.05)")
```

## Enrichment exercises

1. Run enrichment of the current significant gene list against the canonical pathways `C2.CP` database. Do you get more or fewer significant gene sets? Is this what you'd expect give what you know about C2?
1. Run Hallmark enrichment on genes significant at FDR < 0.01. Do you get more or fewer significant gene sets? Why might this be the case?
1. Further modify the final ggplot. Add color, change the FDR cutoff, do whatever your heart desires!

# Gene set enrichment analysis (GSEA)


# Additional resources
## Groups

* [Rladies Seattle](https://www.meetup.com/rladies-seattle/) Not just for ladies! A pro-actively inclusive R community with both in-person and online workshops, hangouts, etc.
* [TidyTuesday](https://github.com/rfordatascience/tidytuesday) A weekly plotting challenge
* [R code club](https://www.riffomonas.org/code_club/) Dr. Pat Schloss opened his lab's coding club to remote participation. 
* [Seattle useR Group](https://www.meetup.com/Seattle-useR/)

## Online

* [R cheatsheets](https://www.rstudio.com/resources/cheatsheets/) also available in RStudio under Help > Cheatsheets
* [The Carpentries](https://carpentries.org/)

# R session

```{r}
sessionInfo()
```

***